File: 152. Maximum Product Subarray.cpp
class Solution {
public:
    /**
     * @param nums: a vector of integers
     * @return: an integer
     */
    int maxProduct(vector<int>& nums) {
        // write your code here
        int pos_max = nums[0];
        int neg_max = nums[0];
        int ret = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            int tmp_pos_max = pos_max;
            int tmp_neg_max = neg_max;
            pos_max = max(nums[i], max(nums[i]*tmp_pos_max, nums[i]*tmp_neg_max));
            neg_max = min(nums[i], min(nums[i]*tmp_pos_max, nums[i]*tmp_neg_max));
            ret = max(pos_max, ret);
        }
        
        return ret;
    }
};

File: go/src/leetcode/152. Maximum Product Subarray/152. Maximum Product Subarray.go
package main

import (
	"math"
)

func maxProduct(nums []int) int {
	// f1保存当前连续数组的最大值, f2保存当前连续数组的最小值
	f1, f2 := 1, 1
	maxVal := math.MinInt32;
	for _, num := range nums {
		if num < 0 {
			f1, f2 = f2, f1
		}
		f1 = max(num, f1*num)
		f2 = min(num, f2*num)

		maxVal = max(maxVal, f1)
	}

	return maxVal
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
